### PR TITLE
feat: Option.or

### PR DESCRIPTION
**Pull request by TwoFX**
Split from #4583

### CODE REVIEW

File: src/Init/Data/Option/Basic.lean [hunk 1]
```diff
@@ -89,6 +89,11 @@ Implementation of `OrElse`'s `<|>` syntax for `Option`.
 instance : OrElse (Option α) where
   orElse := Option.orElse
 
+/-- If the first argument is `some a`, returns `some a`, otherwise returns the second argument. -/
+@[always_inline, inline] def or : Option α → Option α → Option α
```
Comment (digama0) [inline src/Init/Data/Option/Basic.lean:93]: should this be `macro_inline`? That way the second argument will not be evaluated if not needed, similar to `cond`

File: src/Init/Data/Option/Basic.lean [hunk 2]
```diff
@@ -89,6 +89,11 @@ Implementation of `OrElse`'s `<|>` syntax for `Option`.
 instance : OrElse (Option α) where
   orElse := Option.orElse
 
+/-- If the first argument is `some a`, returns `some a`, otherwise returns the second argument. -/
```
Comment (semorrison) [inline src/Init/Data/Option/Basic.lean:92]: Could the doc-string explicitly reference `orElse` (and vice versa). I worry still about people being confused about which one they should be using.

File: src/Init/Data/Option/Basic.lean (hunk 1 above)
Comment (TwoFX) [reply src/Init/Data/Option/Basic.lean:96]: I think if this behavior is desired then `orElse` is a better choice. I have updated the docstring to say as much.

Comment (digama0) [reply src/Init/Data/Option/Basic.lean:96]: I agree, but it still seems unnecessary and inconsistent to have eager evaluation behavior when all other similar functions (like `Bool.or`) have macro inline behavior.

Comment (TwoFX) [reply src/Init/Data/Option/Basic.lean]: I guess we don't lose anything by making it `macro_inline`. Done.

**Pull request merged**